agents/ahv/fence_ahv.py

```diff
@@ -0,0 +1,755 @@
+        elif filter_str and not limit:
+            vm_url = f"{vm_url}?$filter={filter_str}"
+        elif limit and not filter_str:
+            vm_url = f"{vm_url}?$limit={limit}"
+
+        logging.info("Sending GET request to get VM details, %s", vm_url)
+        header_str = self._get_headers()
+
+        try:
+            resp = self.request(url=vm_url, method='GET',
+                                headers=header_str, verify=self.verify)
+        except NutanixClientException as err:
+            logging.error("Unable to retrieve VM info")
+            raise AHVFenceAgentException from err
+
+        vms = resp.json()
+        return vms
+
+    def _get_vm_uuid(self, vm_name):
+        """
+        Get VM UUID given a VM name
+
+        Args:
+            vm_name(str): VM name
+
+        Returns:
+            str: VM UUID
+        """
+        vm_uuid = None
+        resp = None
+
+        if not vm_name:
+            logging.error("VM name was not provided")
+            raise AHVFenceAgentException("VM name not provided")
+
+        try:
+            filter_str = f"name eq '{vm_name}'"
+            resp = self._get_all_vms(filter_str=filter_str)
+        except AHVFenceAgentException as err:
+            logging.error("Failed to get VM info for VM %s", vm_name)
+            raise AHVFenceAgentException from err
+
+        for vm in resp['data']:
+            if vm['name'] == vm_name:
+                vm_uuid = vm['extId']
+                break
+
+        return vm_uuid
+
+    def _get_vm(self, vm_uuid):
+        """
+        Get config details of an existing VM.
+
+        Args:
+            vm_uuid(str): VM UUID.
+
+        Returns:
+            object: HTTP response object.
+        """
+        if not vm_uuid:
+            logging.error("VM UUID was not provided")
+            raise AHVFenceAgentException("VM UUID not provided")
+
+        vm_url = self.vm_url + f"/{vm_uuid}"
+        logging.info("Sending GET request to get VM details, %s", vm_url)
+        header_str = self._get_headers()
+
+        try:
+            resp = self.request(url=vm_url, method='GET',
+                                headers=header_str, verify=self.verify)
+        except NutanixClientException as err:
+            logging.error("Failed to retrieve VM details "
+                          "for VM UUID: vm_uuid")
+            raise AHVFenceAgentException from err
+
+        return resp
+
+    def _power_off_vm(self, vm_uuid):
+        """
+        Force hard power off of a VM
+
+        Args:
+            vm_uuid(str): VM UUID string
+
+        Returns:
+            object: HTTP response object.
+        """
+        resp = None
+
+        if not vm_uuid:
+            logging.error("VM UUID was not provided")
+            raise AHVFenceAgentException("VM UUID not provided")
+
+        vm_url = self.vm_url + f"/{vm_uuid}/$actions/power-off"
+        logging.info("Sending request to power off VM, %s", vm_url)
```

Comment:
Nit: I would recommend to use the vm_uuid here. perhaps in the request method you can log the actual URL which is being sent to the PC. but you can leav ethat one as log.debug not info. This way we won't flood the logs with every API call!

Reply:
ditto for the rest of the debug messages

Reply:
Done